linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,247 @@
+                let mut seeds = Vec::new();
+                get_seeds(n.clone(), &neighbors, &mut points, &processed, &mut seeds);
+                while !seeds.is_empty() {
+                    seeds.sort_unstable_by(|a, b| b.cmp(a));
+                    let (i, min_point) = seeds
+                        .iter()
+                        .enumerate()
+                        .min_by(|(_, a), (_, b)| points[**a].cmp(&points[**b]))
+                        .unwrap();
+                    let n = &mut points[*min_point];
+                    seeds.remove(i);
+                    processed.insert(n.index);
+                    let neighbors = find_neighbors(&n.observation, observations, self.tolerance());
+
+                    n.set_core_distance(self.minimum_points(), &neighbors);
+                    result.orderings.push(n.sample());
+                    if n.c_distance.is_some() {
+                        get_seeds(n.clone(), &neighbors, &mut points, &processed, &mut seeds);
+                    }
+                }
+            }
+        }
+        result
+    }
+}
+
+/// For a sample find the points which are directly density reachable which have not
+/// yet been processed
+fn get_seeds<'a, F: Float>(
+    sample: Neighbor<'a, F>,
+    neighbors: &[Neighbor<'a, F>],
+    points: &mut [Neighbor<'a, F>],
+    processed: &HashSet<usize>,
+    seeds: &mut Vec<usize>,
+) {
+    for n in neighbors.iter().filter(|x| !processed.contains(&x.index)) {
+        let dist = FloatOrd(n.observation.l2_dist(&sample.observation).unwrap());
+        let r_dist = max(sample.c_distance.unwrap(), dist);
+        match points[n.index].r_distance {
+            None => {
+                points[n.index].r_distance = Some(r_dist);
+                seeds.push(n.index);
+            }
+            Some(s) if r_dist < s => points[n.index].r_distance = Some(r_dist),
+            _ => {}
+        }
+    }
+}
+
+/// Given a candidate point, a list of observations, epsilon and list of already
+/// assigned cluster IDs return a list of observations that neighbor the candidate. This function
+/// uses euclidean distance.
+fn find_neighbors<'a, F: Float>(
+    candidate: &ArrayBase<impl Data<Elem = F>, Ix1>,
+    observations: &'a ArrayBase<impl Data<Elem = F>, Ix2>,
+    eps: f64,
+) -> Vec<Neighbor<'a, F>> {
+    let mut res = Vec::new();
+    for (i, obs) in observations.axis_iter(Axis(0)).enumerate() {
+        let distance = candidate.l2_dist(&obs).unwrap();
+        if distance < eps {
+            res.push(Neighbor {
+                index: i,
+                processed: false,
+                observation: obs,
+                r_distance: Some(FloatOrd(distance)),
+                c_distance: None,
+            });
+        }
+    }
+    res.sort_unstable();
+    res
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ndarray::Array2;
+
+    #[test]
+    fn simple_optics() {
```

Comment:
you can add tests for the following categories:
 * test single functions, i.e. check that `find_neighbors` and `get_seeds` are working as intended
 * do qualitative testing, i.e. construct simple example where the ordering is clear, probably with only a handful of points 
 * do quantitative testing, i.e. generate random data and prove effectiveness in separating them

Reply:
I need to do some testing for `get_seeds` but I've added some consistency tests to make sure certain properties are kept with the returned values. I'll add some more tests tomorrow if I have time but it looks like it's coming along :grin: